states/module.go

```diff
@@ -88,23 +88,59 @@ func (ms *Module) RemoveResource(addr addrs.Resource) {
 // are updated for all other instances of the same resource as a side-effect of
 // this call.
 func (ms *Module) SetResourceInstanceCurrent(addr addrs.ResourceInstance, obj *ResourceInstanceObjectSrc, provider addrs.AbsProviderConfig) {
-	ms.SetResourceMeta(addr.Resource, eachModeForInstanceKey(addr.Key), provider)
-
 	rs := ms.Resource(addr.Resource)
-	is := rs.EnsureInstance(addr.Key)
-
-	is.Current = obj
-
-	if !is.HasObjects() {
-		// If we have no objects at all then we'll clean up.
-		delete(rs.Instances, addr.Key)
+	// if the resource is nil and the object is nil, don't do anything!
+	// you'll probably just cause issues
+	if obj == nil && rs == nil {
+		return
 	}
-	if rs.EachMode == NoEach && len(rs.Instances) == 0 {
-		// Also clean up if we only expect to have one instance anyway
-		// and there are none. We leave the resource behind if an each mode
-		// is active because an empty list or map of instances is a valid state.
-		delete(ms.Resources, addr.Resource.String())
+	if obj == nil && rs != nil {
+		// does the resource have any other objects?
+		// if not then delete the whole resource
+		// When deleting the resource, ensure that its EachMode is NoEach,
+		// as a resource with EachList or EachMap can have 0 instances and be valid
+		if rs.EachMode == NoEach && len(rs.Instances) == 0 {
+			delete(ms.Resources, addr.Resource.String())
+			return
+		}
+		// check for an existing resource, now that we've ensured that rs.Instances is more than 0/not nil
+		is := rs.Instance(addr.Key)
+		if is == nil {
+			// if there is no instance, but the resource exists and has other instances,
```

Comment:
I think this comment is now not totally accurate, because we can get here in the case where the resource exists but has _no_ other instances. The code around it seems correct, but since you've gone to the effort of adding all of these great comments I thought you'd probably want to clarify this one. :grinning: